### DISCUSSION TITLE
Expected survival time

### DISCUSSION THREAD

**Discussion by cedricdeboom** [CamDavidsonPilon/lifelines#1318] [Ideas]
It would be nice to be able to compute the expected survival time from a fitted survival function. It should not be that hard actually, see: https://math.stackexchange.com/questions/642021/expected-value-as-integral-of-survival-function

Now I always do it manually, but it would make a nice addition to Lifelines!

**Comment by CamDavidsonPilon** [comment]
Agree! We currently have `median_survival_time_`, so a `mean_survival_time_` makes sense too. I'll convert this to an issue.

**Reply by josegalelop1345** [reply]
Hey @cedricdeboom @CamDavidsonPilon  , can I kindly ask you how do you calculate the integral of the survival function?
As of now, when predicting for my population, I obtain a bunch of arrays with the survival function for each time period.

I guess I need to calculate the integral by an approximation method, but I'm not fluent enough in python/math to know how to do so..
Thanks in advance!!

**Reply by josegalelop1345** [reply]
PS: My goal is to compute the expected survival time from a fitted survival function

**Reply by cedricdeboom** [reply]
I did it like this, but I have not tested nor rigorously proven that this is a correct method. It could be that I am forgetting some factors, but my goal was to compare different models to each other, rather than have a near 100% correct absolute expected value. This is the code:
```
# Get the survival function
sf = kmf_full.survival_function_

# Calculate the approximate integral
integral = 0.0
previous_t = 0.0
previous_value = 1.0
for i, (j, r) in enumerate(sf.iterrows()):
    value = r['KM_estimate']
    t = r.name
    integral += previous_value * (t - previous_t)
    previous_t = t
    previous_value = value

integral += previous_value * 1.0
print(f"Expected survival time: {integral}")
```

**Reply by josegalelop1345** [reply]
@cedricdeboom Many thanks for this!

**Reply by CamDavidsonPilon** [reply]
I would suggest looking into `scipy.integrate.trapezoid`: https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.trapezoid.html

```python
ev = trapezoid(km.survival_function_['KM_estimate'].values, km.survival_function_.index)
```

**Reply by cedricdeboom** [reply]
Even better!
